gems/sorbet-runtime/lib/types/types/base.rb

```diff
@@ -161,8 +161,12 @@ def hash
     end
 
     def ==(other)
-      (T::Utils.resolve_alias(other).class == T::Utils.resolve_alias(self).class) &&
+      case other
```

Comment:
This change preserves equality between `T::Types::Union` and `T::Private::Types::SimplePairUnion` when they're equivalent (which can basically only happen if the former is explicitly constructed, circumventing `T.any`).

It's slightly wider-radius than needed - we could just do this on `T::Types::Union` - but this is a more standard way of implementing equality and will only produce different results if someone makes a class whose name makes it pretend to be a non-simple `T::Types::Base`, which doesn't seem worth worrying about (seems in the same category as overriding `class` for the current implementation).